### PR TITLE
feat: read team token from kafka headers in blobby

### PR DESCRIPTION
**Pull request by pauldambra**
Recently we read tens of millions of messages knowing the only action was to drop them. For every message we had to parse the body to get the token so we could check if we wanted to drop it.

Let's put the team token into the kafka headers since if there's an incident we always care about those values.

And then we can read the team token from the header and drop the message without parsing it when the team does not have session recording enabled.

This still _also_ reads the body to check for team token as a fallback. If nothing else this lets us deploy this without worrying about rolling out a switcheroo

So

token in headers, team not found - return early (new behaviour)
token in headers, team found - does parse the body, but doesn't check the team again
token not in headers, team not found - return early (as before)
token not in headers, team found - (as before)

### CODE REVIEW

File: plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer.ts [hunk 1]
```diff
@@ -106,6 +106,38 @@ export interface TeamIDWithConfig {
...
+
+    if (token) {
+        const teams = await teamsRefresher.get()
+        teamIdWithConfig = {
+            teamId: teams[token]?.teamId || null,
+            consoleLogIngestionEnabled: teams[token]?.consoleLogIngestionEnabled ?? true,
+        }
+    }
+    return { token, teamIdWithConfig }
+}
```
Comment (benjackwhite) [inline plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer.ts:139]: Could we move this to utils?

File: plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer.ts [hunk 2]
```diff
@@ -422,7 +457,23 @@ export class SessionRecordingIngester {
...
+                                    .labels({
+                                        event_type: 'session_recordings_blob_ingestion',
+                                        drop_cause: 'team_missing_or_disabled',
+                                    })
+                                    .inc()
+
+                                return statusWarn('team_missing_or_disabled', {
+                                    token: token,
+                                })
+                            }
```
Comment (benjackwhite) [inline plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer.ts:474]: Could we not do this in `parseKafkaMessage`? Feels like it fits better inside the function

**Review by benjackwhite — CHANGES_REQUESTED** [review]

File: plugin-server/src/main/ingestion-queues/session-recording/services/session-manager.ts [hunk 3]
```diff
@@ -217,7 +217,7 @@ export class SessionManager {
             return
         }
 
-        const lagMultiplier = getLagMultipler(partitionLag)
+        const lagMultiplier = getLagMultiplier(partitionLag)
```
Comment (benjackwhite) [inline plugin-server/src/main/ingestion-queues/session-recording/services/session-manager.ts:220]: Multipler is the formal version of the word... 🙈

**Review by benjackwhite — APPROVED** [review]
This looks great. Was going to comment on how will we know if it works but i guess thats the distinction of the `drop_cause` ?

**Comment by pauldambra** [comment]
> Was going to comment on how will we know if it works but i guess thats the distinction of the drop_cause

yep, I made that more explicit so we'll see no more of the existing drop cause and one or other of the new ones 👍

File: plugin-server/src/main/ingestion-queues/session-recording/utils.ts [hunk 4]
```diff
@@ -102,10 +105,169 @@ export const getPartitionsForTopic = (
...
+        event = JSON.parse(messagePayload.data)
+    } catch (error) {
+        return statusWarn('invalid_json', { error })
+    }
+
+    const { $snapshot_items, $session_id, $window_id } = event.properties || {}
+
+    // NOTE: This is simple validation - ideally we should do proper schema based validation
+    if (event.event !== '$snapshot_items' || !$snapshot_items || !$session_id) {
+        status.warn('🙈', 'Received non-snapshot message, ignoring')
```
Comment (xvello) [inline plugin-server/src/main/ingestion-queues/session-recording/utils.ts:189]: let's increment `eventDroppedCounter` too here for visibility. Ideally, I'd love for all drop scenarios to increment this metric.

File: plugin-server/src/main/ingestion-queues/session-recording/utils.ts [hunk 5]
```diff
@@ -102,10 +105,169 @@ export const getPartitionsForTopic = (
...
+        eventDroppedCounter
+            .labels({
+                event_type: 'session_recordings_blob_ingestion',
+                drop_cause: 'header_token_team_missing_or_disabled',
+            })
+            .inc()
+
+        return statusWarn('header_token_team_missing_or_disabled', {
+            token: token,
+        })
```
Comment (xvello) [inline plugin-server/src/main/ingestion-queues/session-recording/utils.ts:172]: We have an opportunity to drop here if `!teamIdWithConfig.session_recording_opt_in`, don't we?

**Review by xvello — APPROVED** [review]

File: plugin-server/src/main/ingestion-queues/session-recording/utils.ts (hunk 5 above)
Comment (pauldambra) [reply plugin-server/src/main/ingestion-queues/session-recording/utils.ts:172]: That's implicit in this check. We only load teams that exist _and_ have `session_recording_opt_in` set to true.

I guess the assumption was it wasn't valuable enough to know the difference in the event dropped counter to stand the cost of loading more teams...

**Review by benjackwhite — APPROVED** [review]

**Pull request merged**